Wasm: improve debug ux (#4534)
JetBrains/kotlin-web-site: docs/topics/wasm/wasm-debugging.md

3. Select the **Web** option. Make sure that no other options are selected.
4. Click the **Download** button and unpack the resulting archive.

> ![Kotlin Multiplatform wizard](wasm-compose-web-wizard.png){width=600}

Reviewer: I think you can cut the width in half, now the wizard seems too big compared to the instruction list

The same is true for all other images on this page that have big lengths. The maximum width makes them too big, taking up the whole page scroll, like in the Custom formatters section

Author: Adjusted all images in a size that it's not too big but allows clear visibility.

   * ![Step over](wasm-step-over.png){width=30}{type="joined"} Step over to execute the current line and pause on the next line.
   * ![Step out](wasm-step-out.png){width=30}{type="joined"} Step out to execute the code until it exits the current function.

   ![Debug controls](wasm-debug-controls.png){width=700}

6. Check the **Call stack** and **Scope** panes to trace the sequence of function calls and pinpoint the location of any errors.

   ![Check call stack](wasm-debug-scope.png){width=700}

   For an improved visualization of the variable values, see [Enhance your debugging with custom formatters](#enhance-your-debugging-with-custom-formatters).

7. Make changes to your code and [run the application](#run-the-application) again to verify that everything works as expected.
8. Click on the line numbers with breakpoints to remove the breakpoints.

### Enhance your debugging with custom formatters

> In addition to the default debugging experience, you can set up custom formatters to display and locate variable values in a more user-friendly and comprehensible manner.

Reviewer: ```suggestion
In addition to the default debugging configuration, you can set up custom formatters to display and locate variable values in a more user-friendly and comprehensible manner.
```

![Kotlin/Wasm improved debugger](wasm-debugger-improved.png){width=700}

This implementation is supported across major browsers like Firefox and Chromium-based as
it uses the [custom formatters API](https://firefox-source-docs.mozilla.org/devtools-user/custom_formatters/index.html).

To set up custom formatters for an improved debugging experience:

1. Add the following compiler argument to the `wasmJs` compiler options:

> ```kotlin

Reviewer: The code block is not aligned with the first item

kotlin {
    wasmJs {

>         …

Reviewer: ```suggestion
        // ...
```

Reviewer: I think we usually write it like this to avoid red code when copy and pasting

        compilerOptions {
            freeCompilerArgs.add("-Xwasm-debugger-custom-formatters")
        }
    }
}   
```

> 2. Enable the **Custom formatters** feature in your browser.

Reviewer: ```suggestion
2. Enable the **Custom formatters** feature in your browser:
```

>    In the Chrome DevTools, it’s placed in **Settings | Preferences | Console**:

Reviewer: ```suggestion
   In the Chrome DevTools, it's placed in **Settings | Preferences | Console**:
```

>    In the Chrome DevTools, it’s placed in **Settings | Preferences | Console**:

Reviewer: This could also be a list, for example:

2. Enable the **Custom formatters** feature in your browser:
    * In the Chrome DevTools,
    * In the Firefox Developer Tools,

   ![Enable custom formatters in Chrome](wasm-custom-formatters-chrome.png){width=700}

>    In the Firefox Developer Tools, it’s placed in **Settings | Advanced settings**:

Reviewer: ```suggestion
   In the Firefox Developer Tools, it's placed in **Settings | Advanced settings**:
```

   ![Enable custom formatters in Firefox](wasm-custom-formatters-firefox.png){width=700}

> After setting up the custom formatters, you can continue the debugging steps explained in

Reviewer: ```suggestion
After setting up custom formatters, you can complete the
```

> After setting up the custom formatters, you can continue the debugging steps explained in

Reviewer: I think it would be great to make the flow of the whole page more linear so that users don't have to come back and redo the steps in Debug your Kotlin/Wasm application after new browser setup

Did you consider adding this info to the beginning, for example, the Configure your browser for debugging section?

> [Debug your Kotlin/Wasm application](#debug-your-kotlin-wasm-application).

Reviewer: ```suggestion
[Debug your Kotlin/Wasm application](#debug-your-kotlin-wasm-application) tutorial.
```
